README.md
# Multiselect Search

Easily buildable and customizable multi select widget with search field that can be used anywhere from container to dialog

![ss3](https://user-images.githubusercontent.com/34756693/198015842-83e5f7b1-dc55-4122-8368-2af76b0ac459.gif)
![ss2](https://user-images.githubusercontent.com/34756693/198015830-d7f9d45e-f8c4-44b7-985b-3825efe9b14f.gif)
![ss1](https://user-images.githubusercontent.com/34756693/198015847-155346a3-1a48-495a-8792-88d0975c066b.gif)


## Features

- Can be used in any Flutter widget. Wrap it in Container, dialog, bottom modal sheet etc.
- User can search from your list of items.
- Selected item chips list, search field, selectable item list are all visible in one.
- List items are not hard-coded. You make your own list item widgets.
- `onChanged` method returns a list of selected items every time user selects/unselect an item.

## Getting started

To start using the package, add the dependencies in your pubspec.yaml and import.

## Usage

1. First, you need a list of items from same Class.
   Your class model **MUST** have `fromJson` & `toJson` method.

    ```dart
    var list = [
        Contact(1, "Joel McHale"),
        Contact(2, "Danny Pudi"),
        Contact(3, "Donald Glover"),
        Contact(4, "Gillian Jacobs"),
        Contact(5, "Alison Brie"),
        Contact(6, "Chevy Chase"),
        Contact(7, "Jim Rush"),
        Contact(8, "Yvette Nicole Brown"),
        Contact(9, "Jeff Winger"),
        Contact(10, "Abed Nadir"),
        Contact(11, "Troy Barnes"),
        Contact(12, "Britta Perry"),
        Contact(13, "Annie Edison"),
        ];

    class Contact {
        final int id;
        final String name;

        Contact(
            this.id,
            this.name,
        );

        Contact.fromJson(Map<String, dynamic> json)
            : id = json['id'],
                name = json['name'];

        Map<String, dynamic> toJson() {
                return {
                'id': id,
                'name': name,
                };
            }
    }
    ```

2. Now you are ready to use the `MultiSelectSearch` widget.
   - `itemBuilder`
    : Build your own list item widgets to select from.

      ![itemBuilder](https://user-images.githubusercontent.com/34756693/198016065-dee0b226-0078-435d-850c-23ef75e3e7ef.png)

   - `chipLabelKey`
    : A field name of your class that should be displayed in selected item chip. In this example, we're displaying Contact's name field value in chip when selected. So it's `'name'`

   - `items`
    : The list you created in step 1. Both items and itemBuilder must use the same list.

   - `initialValue`
    : Some of your list items to be selected beforehand.

   - `onChanged`
    : Returns the selected item list everytime a user selects/unselects an item.

   - `clearAll`
    : A widget to clear all selected items when clicked. Do not use button widgets.

      ![clear](https://user-images.githubusercontent.com/34756693/198016125-c547d3df-26c3-41da-b672-73e98b14df40.png)

        ```dart
        MultiSelectSearch<Contact>(
            itemBuilder: (Contact item) => ListTile(
                key: ObjectKey(item),
                leading: const Icon(Icons.person),
                title: Text(item.name),
            ),
            chipLabelKey: 'name',
            items: list,
            initialValue: initial,
            onChanged: (List<Contact> items) =>
                setState(() => selectedItems = items),
            decoration: BoxDecoration(
                color: const Color(0xFFF7A072).withOpacity(0.6),
                border: const Border(
                    bottom: BorderSide(color: Colors.grey),
                ),
            ),
            clearAll: const Padding(
                padding: EdgeInsets.only(top: 10.0, right: 6.0),
                child: Icon(Icons.clear),
            ),
        ),
        ```

3. To customize some styles:

   - `maxHeight`: Search field & selected items container's maximum height when it's expanded.

   - `decoration`: Search field & selected items container decoration

   - `padding`: Search field & selected items container padding
  
   - `searchFieldDecoration`: `InputDecoration` to change search field's style, hint text and more.

       ![style1](https://user-images.githubusercontent.com/34756693/198016208-3e0578ce-aa70-4272-9942-a52814d266ad.png)

   - Selected item chip gets its style from the theme. In order to customize it, change `chipThemeData` in your AppTheme

        ```dart
        theme: ThemeData(
                primarySwatch: Colors.teal,
                chipTheme: const ChipThemeData(
                deleteIconColor: Color.fromARGB(255, 61, 61, 61),
                backgroundColor: Color(0xFFF9F7F3),
                ),
            ),
        ```

        ![chip](https://user-images.githubusercontent.com/34756693/198016260-388fba21-3278-495a-9d5e-d952b0377b5f.png)

## Additional information

Feel free to create issue for any bugs or requests.


CHANGELOG.md
## 0.0.1 2022/10/26

- Initially published multi_select_search package.
- Includes example app.
- Widgets: MultiSelectSearch, SearchField.
## 0.0.2 2022/10/26

- Updated gifs & images in README
## 0.0.3 2022/10/27

- README fix


test/multi_select_search_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:multi_select_search/multi_select_search.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/search_field.dart
import 'package:flutter/material.dart';

/// A text field to search item from the list
class SearchField extends StatelessWidget {
  const SearchField({
    Key? key,
    this.onTap,
    this.stepWidth,
    this.onChanged,
    this.decoration,
    required this.focusNode,
    required this.controller,
  }) : super(key: key);

  final double? stepWidth;
  final FocusNode focusNode;
  final GestureTapCallback? onTap;
  final InputDecoration? decoration;
  final ValueChanged<String>? onChanged;
  final TextEditingController controller;

  @override
  Widget build(BuildContext context) {
    return IntrinsicWidth(
      stepWidth: stepWidth,
      child: TextField(
        scrollPadding: EdgeInsets.zero,
        keyboardType: TextInputType.multiline,
        minLines: 1,
        maxLines: 2,
        textAlignVertical: TextAlignVertical.center,
        controller: controller,
        onChanged: (newValue) {
          onChanged!(newValue);
        },
        focusNode: focusNode,
        onEditingComplete: () => FocusScope.of(context).nextFocus(),
        onTap: onTap,
        scrollPhysics: const NeverScrollableScrollPhysics(),
        decoration: decoration ??
            const InputDecoration(
              hintText: 'Search here..',
              border: InputBorder.none,
            ),
        cursorHeight: 20,
      ),
    );
  }
}


lib/src/multi_select_search.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:multi_select_search/src/search_field.dart';

/// Make sure your model class must contain fromJson & toJson method/
/// If you want to change selected item [Chip] theme change [chipThemeData] in your AppTheme
class MultiSelectSearch<T> extends StatefulWidget {
  /// A builder to display list of widgets that you can select/unselect from.
  final Widget Function(T) itemBuilder;

  /// A field name of your model to show its value in [Chip]
  /// For examle: You have a List<Contact> and when a user selects from the list
  /// you want to show the selected contact's name in the chip.  e.g.Contact.name
  /// In this situation, you define chipLabelKey: 'name'
  final String chipLabelKey;

  /// List of all items to select from
  final List<T> items;

  /// The list of initial selected items
  final List<T> initialValue;

  /// Search field & selected items [Container]'s max height
  final double? maxHeight;

  /// Search field & selected items [Container] decoration
  final Decoration? decoration;

  /// Search field & selected items [Container] padding
  final EdgeInsetsGeometry? padding;

  /// Called everytime when user selects or unselects an item.
  /// Returns selected items.
  /// If user selects an item it gets added to the selected items.
  /// If user unselect by tapping on deleteIcon on the chip, it gets removed from the selected items.
  final void Function(List<T> data) onChanged;

  /// Search field input decoration to change its style, hint text and more.
  final InputDecoration? searchFieldDecoration;

  // Chip related
  /// Selected item [Chip]'s text style
  final TextStyle? chipTextStyle;

  /// A widget to display when there is no more item to choose in the list
  final Widget? emptyListIndicator;

  /// When clicked, all selected items will be cleared.
  /// Don't give button widgets because they have their own onPressed method.
  final Widget? clearAll;

  const MultiSelectSearch({
    Key? key,
    this.padding,
    this.clearAll,
    this.maxHeight,
    this.decoration,
    this.chipTextStyle,
    required this.items,
    required this.onChanged,
    this.emptyListIndicator,
    required this.itemBuilder,
    required this.initialValue,
    this.searchFieldDecoration,
    required this.chipLabelKey,
  }) : super(key: key);

  @override
  State<MultiSelectSearch<T>> createState() => _MultiSelectSearchState<T>();
}

class _MultiSelectSearchState<T> extends State<MultiSelectSearch<T>> {
  final _contactListScrollController = ScrollController();
  final _chipsScrollController = ScrollController();
  final _controller = TextEditingController();
  final _focusNode = FocusNode();

  late final ValueNotifier<List<T>> _listViewItems;
  List<T> _selectedItems = [];
  List<T> _availableItems = [];

  @override
  void initState() {
    super.initState();
    try {
      jsonDecode(jsonEncode(widget.items.first));
    } on JsonUnsupportedObjectError {
      throw FlutterError.fromParts(<DiagnosticsNode>[
        ErrorSummary(
            '${widget.items.first.runtimeType} must contain fromJson and toJson method.'),
      ]);
    }
    _selectedItems = [...widget.initialValue];
    _listViewItems = ValueNotifier(_removeSelectedItems());
    _availableItems = [..._listViewItems.value];
    _controller.addListener(() => _searchFilter());
  }

  @override
  void dispose() {
    _focusNode.dispose();
    _chipsScrollController.dispose();
    _contactListScrollController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final chips = List<Widget>.generate(_selectedItems.length, (index) {
      final label =
          jsonDecode(jsonEncode(_selectedItems[index]))[widget.chipLabelKey];
      return Padding(
        padding: const EdgeInsets.only(top: 10.0),
        child: Chip(
          label: Text(label, style: widget.chipTextStyle),
          labelPadding:
              const EdgeInsets.symmetric(horizontal: 4.0, vertical: 0.0),
          onDeleted: () => _deleteChip(_selectedItems[index]),
        ),
      );
    });

    chips.add(
      SearchField(
        stepWidth: 100,
        controller: _controller,
        focusNode: _focusNode,
        decoration: widget.searchFieldDecoration,
      ),
    );

    return ValueListenableBuilder(
      valueListenable: _listViewItems,
      builder: ((context, listItems, child) {
        return Column(
          mainAxisAlignment: MainAxisAlignment.start,
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(
              constraints: BoxConstraints(maxHeight: widget.maxHeight ?? 100.0),
              padding: widget.padding ?? const EdgeInsets.only(left: 8.0),
              decoration: widget.decoration,
              child: Row(
                crossAxisAlignment: CrossAxisAlignment.start,
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Expanded(
                    child: GestureDetector(
                      behavior: HitTestBehavior.translucent,
                      onTap: () =>
                          FocusScope.of(context).requestFocus(_focusNode),
                      child: SingleChildScrollView(
                        scrollDirection: Axis.vertical,
                        controller: _chipsScrollController,
                        child: Padding(
                          padding: const EdgeInsets.only(right: 16.0),
                          child: Wrap(
                            spacing: 6.0,
                            alignment: WrapAlignment.start,
                            children: chips,
                          ),
                        ),
                      ),
                    ),
                  ),
                  if (widget.clearAll != null)
                    MouseRegion(
                      cursor: SystemMouseCursors.click,
                      child: GestureDetector(
                        behavior: HitTestBehavior.translucent,
                        onTap: () => _clearAll(),
                        child: widget.clearAll,
                      ),
                    )
                ],
              ),
            ),
            // Item list view
            Flexible(
              child: listItems.isEmpty
                  ? widget.emptyListIndicator ??
                      const Padding(
                        padding: EdgeInsets.all(16.0),
                        child: Text("No more items"),
                      )
                  : ListView.builder(
                      shrinkWrap: true,
                      controller: _contactListScrollController,
                      itemCount: listItems.length,
                      itemBuilder: (BuildContext context, int index) {
                        return InkWell(
                          onTap: () => _selectItem(listItems[index]),
                          child: widget.itemBuilder(listItems[index]),
                        );
                      },
                    ),
            ),
          ],
        );
      }),
    );
  }

  void _clearAll() {
    setState(() => _selectedItems = []);
    _controller.clear();
    _listViewItems.value = [...widget.items];
    _availableItems = [...widget.items];
    widget.onChanged.call(_selectedItems);
  }

  void _deleteChip(T item) {
    setState(() => _selectedItems.remove(item));
    _listViewItems.value.insert(0, item);
    _availableItems.insert(0, item);

    if (_controller.text.isNotEmpty) _searchFilter();
    widget.onChanged.call(_selectedItems);
  }

  void _selectItem(T item) {
    _controller.clear();
    _selectedItems.add(item);
    _listViewItems.value.remove(item);
    _availableItems.remove(item);
    FocusScope.of(context).requestFocus(_focusNode);
    widget.onChanged(_selectedItems.cast<T>());
  }

  List<T> _removeSelectedItems() {
    final items = [...widget.items];
    for (var i = 0; i < _selectedItems.length; i++) {
      items.remove(_selectedItems[i]);
    }
    return items;
  }

  void _searchFilter() {
    List<T> items = [..._availableItems];

    final itemsMap = jsonDecode(jsonEncode(items));

    List<T> filteredItems = [];
    for (var i = 0; i < _availableItems.length; i++) {
      if (itemsMap[i][widget.chipLabelKey]
          .toString()
          .toLowerCase()
          .contains(_controller.text.toLowerCase())) {
        filteredItems.add(_availableItems[i]);
      }
    }

    _listViewItems.value = filteredItems;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:multi_select_search/multi_select_search.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Multiselect Search',
      theme: ThemeData(
        primarySwatch: Colors.teal,
        chipTheme: const ChipThemeData(
          deleteIconColor: Color.fromARGB(255, 61, 61, 61),
          backgroundColor: Color(0xFFF9F7F3),
        ),
      ),
      home: const MyExampleApp(),
    );
  }
}

class MyExampleApp extends StatefulWidget {
  const MyExampleApp({Key? key}) : super(key: key);

  @override
  State<MyExampleApp> createState() => _MyExampleAppState();
}

class _MyExampleAppState extends State<MyExampleApp> {
  List<Contact> selectedItems = [];
  @override
  Widget build(BuildContext context) {
    var list = [
      Contact(1, "Joel McHale"),
      Contact(2, "Danny Pudi"),
      Contact(3, "Donald Glover"),
      Contact(4, "Gillian Jacobs"),
      Contact(5, "Alison Brie"),
      Contact(6, "Chevy Chase"),
      Contact(7, "Jim Rush"),
      Contact(8, "Yvette Nicole Brown"),
      Contact(9, "Jeff Winger"),
      Contact(10, "Abed Nadir"),
      Contact(11, "Troy Barnes"),
      Contact(12, "Britta Perry"),
      Contact(13, "Annie Edison"),
    ];

    List<Contact> initial = [
      list.first,
      list[1],
      list.last,
    ];

    return Scaffold(
      appBar: AppBar(
        title: const Text("Multi Select Search Menu"),
      ),
      body: Column(
        children: [
          Expanded(
            child: MultiSelectSearch<Contact>(
              itemBuilder: (Contact item) => ListTile(
                key: ObjectKey(item),
                leading: const Icon(Icons.person),
                title: Text(item.name),
              ),
              chipLabelKey: 'name',
              items: list,
              initialValue: initial,
              onChanged: (List<Contact> items) =>
                  setState(() => selectedItems = items),
              decoration: BoxDecoration(
                color: const Color(0xFFF7A072).withOpacity(0.6),
                border: const Border(
                  bottom: BorderSide(color: Colors.grey),
                ),
              ),
              clearAll: const Padding(
                padding: EdgeInsets.only(top: 10.0, right: 6.0),
                child: Icon(Icons.clear),
              ),
            ),
          ),
          Wrap(
            children: [
              for (var i = 0; i < selectedItems.length; i++)
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 8.0),
                  child: Text(selectedItems[i].name),
                )
            ],
          )
        ],
      ),
    );
  }
}

class Contact {
  final int id;
  final String name;

  Contact(
    this.id,
    this.name,
  );

  Contact.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        name = json['name'];

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
    };
  }
}
